RLS always fails `textDocument/definition`, ctrl-hover and ctrl-click no longer working

rls-preview 0.130.5 (1c755ef 2018-10-20) rustc nightly-2018-10-28

Ctrl-click and ctrl-hover no longer work, at all. "result" is always []

```
LSP4E to org.eclipse.corrosion.rls:{"jsonrpc":"2.0","id":"1006","method":"textDocument/definition","params":{"textDocument":{"uri":"***/src/render.rs"},"uri":"***/src/render.rs","position":{"line":92,"character":52}}}
org.eclipse.corrosion.rls to LSP4E:{"jsonrpc":"2.0","id":1006,"result":[]}
```

UPDATE:
- Toolchain nightly-2018-09-20 is unaffected
- Toolchain nightly-2018-10-17 is unaffected
- Toolchain nightly-2018-10-20 is affected
- Toolchain nightly-2018-10-28 is affected
- Toolchain nightly-2018-11-14 is affected
- Toolchain nightly-2018-11-20 is in some cases affected

Just got nightly, and nothing works. All i get is the following

@mickaelistria

```
java.util.concurrent.ExecutionException: org.eclipse.lsp4j.jsonrpc.ResponseErrorException: missing field `codeActionKind`
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1915)
	at org.eclipse.lsp4e.LanguageServerWrapper.getServerCapabilities(LanguageServerWrapper.java:573)
	at org.eclipse.lsp4e.LanguageServiceAccessor.getLSWrappers(LanguageServiceAccessor.java:301)
	at org.eclipse.lsp4e.LanguageServiceAccessor.getLSPDocumentInfosFor(LanguageServiceAccessor.java:444)
	at org.eclipse.lsp4e.ConnectDocumentToLanguageServerSetupParticipant$1.run(ConnectDocumentToLanguageServerSetupParticipant.java:75)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: org.eclipse.lsp4j.jsonrpc.ResponseErrorException: missing field `codeActionKind`
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleResponse(RemoteEndpoint.java:209)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:193)
	at org.eclipse.lsp4e.LanguageServerWrapper.lambda$1(LanguageServerWrapper.java:219)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:192)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:99)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/README.md
+++ b/README.md
@@ -117,7 +117,7 @@ Currently we accept the following options:
   features
 * `racer_completion` (`bool`, defaults to `true`) enables code completion using
   racer (which is, at the moment, our only code completion backend). Also enables
-  hover tooltips to fall back to racer when save-analysis data is unavailable.
+  hover tooltips & go-to-definition to fall back to racer when save-analysis data is unavailable.
 * `clippy_preference` (`String`, defaults to `"opt-in"`) controls eagerness of clippy
   diagnostics when available. Valid values are _(case-insensitive)_:
   - `"off"` Disable clippy lints.

--- a/src/actions/requests.rs
+++ b/src/actions/requests.rs
@@ -24,8 +24,6 @@ use url::Url;
 
 use crate::actions::hover;
 use crate::actions::run::collect_run_actions;
-use crate::actions::work_pool;
-use crate::actions::work_pool::WorkDescription;
 use crate::build::Edition;
 use crate::lsp_data;
 use crate::lsp_data::*;
@@ -203,45 +201,27 @@ impl RequestAction for Definition {
         let span = ctx.convert_pos_to_span(file_path.clone(), params.position);
         let analysis = ctx.analysis.clone();
 
-        // If configured start racer concurrently and fallback to racer result
-        let racer_receiver = {
-            if ctx.config.lock().unwrap().goto_def_racer_fallback {
-                Some(work_pool::receive_from_thread(
-                    move || {
-                        let cache = ctx.racer_cache();
-                        let session = ctx.racer_session(&cache);
-                        let location = pos_to_racer_location(params.position);
-
-                        racer::find_definition(file_path, location, &session)
-                            .and_then(|rm| location_from_racer_match(&rm))
-                    },
-                    WorkDescription("textDocument/definition-racer"),
-                ))
-            } else {
-                None
-            }
-        };
-
         match analysis.goto_def(&span) {
             Ok(out) => {
                 let result = vec![ls_util::rls_to_location(&out)];
                 trace!("goto_def (compiler): {:?}", result);
                 Ok(result)
             }
-            _ => match racer_receiver {
-                Some(receiver) => match receiver.recv() {
-                    Ok(Some(r)) => {
-                        trace!("goto_def (Racer): {:?}", r);
-                        Ok(vec![r])
-                    }
-                    Ok(None) => {
-                        trace!("goto_def (Racer): None");
-                        Ok(vec![])
-                    }
-                    _ => Self::fallback_response(),
-                },
-                _ => Self::fallback_response(),
-            },
+            // If analysis failed & `racer_completion` is enabled try racer
+            _ if ctx.config.lock().unwrap().racer_completion => {
+                let cache = ctx.racer_cache();
+                let session = ctx.racer_session(&cache);
+                let location = pos_to_racer_location(params.position);
+
+                let r = racer::find_definition(file_path, location, &session)
+                    .and_then(|rm| location_from_racer_match(&rm))
+                    .map(|l| vec![l])
+                    .unwrap_or_default();
+
+                trace!("goto_def (Racer): {:?}", r);
+                Ok(r)
+            }
+            _ => Self::fallback_response(),
         }
     }
 }

--- a/src/actions/requests.rs
+++ b/src/actions/requests.rs
@@ -201,14 +201,16 @@ impl RequestAction for Definition {
         let span = ctx.convert_pos_to_span(file_path.clone(), params.position);
         let analysis = ctx.analysis.clone();
 
-        match analysis.goto_def(&span) {
-            Ok(out) => {
-                let result = vec![ls_util::rls_to_location(&out)];
-                trace!("goto_def (compiler): {:?}", result);
-                Ok(result)
-            }
-            // If analysis failed & `racer_completion` is enabled try racer
-            _ if ctx.config.lock().unwrap().racer_completion => {
+        if let Ok(out) = analysis.goto_def(&span) {
+            let result = vec![ls_util::rls_to_location(&out)];
+            trace!("goto_def (compiler): {:?}", result);
+            Ok(result)
+        } else {
+            let racer_enabled = {
+                let config = ctx.config.lock().unwrap();
+                config.racer_completion
+            };
+            if racer_enabled {
                 let cache = ctx.racer_cache();
                 let session = ctx.racer_session(&cache);
                 let location = pos_to_racer_location(params.position);
@@ -220,8 +222,9 @@ impl RequestAction for Definition {
 
                 trace!("goto_def (Racer): {:?}", r);
                 Ok(r)
+            } else {
+                Self::fallback_response()
             }
-            _ => Self::fallback_response(),
         }
     }
 }

--- a/src/build/plan.rs
+++ b/src/build/plan.rs
@@ -53,6 +53,7 @@ crate enum WorkStatus {
     Execute(JobQueue),
 }
 
+#[allow(clippy::large_enum_variant)]
 #[derive(Debug)]
 crate enum BuildPlan {
     External(ExternalPlan),

--- a/src/config.rs
+++ b/src/config.rs
@@ -127,7 +127,6 @@ pub struct Config {
     pub unstable_features: bool,
     pub wait_to_build: Option<u64>,
     pub show_warnings: bool,
-    pub goto_def_racer_fallback: bool,
     /// Clear the RUST_LOG env variable before calling rustc/cargo? Default: true
     pub clear_env_rust_log: bool,
     /// Build the project only when a file got saved and not on file change. Default: false
@@ -140,7 +139,10 @@ pub struct Config {
     pub no_default_features: bool,
     pub jobs: Option<u32>,
     pub all_targets: bool,
-    /// Enable use of racer for `textDocument/completion` requests
+    /// Enable use of racer for `textDocument/completion` requests.
+    ///
+    /// Enabled also enables racer fallbacks for hover & go-to-definition functionality
+    /// if rustc analysis should fail.
     pub racer_completion: bool,
     #[serde(deserialize_with = "deserialize_clippy_preference")]
     pub clippy_preference: ClippyPreference,
@@ -177,7 +179,6 @@ impl Default for Config {
             unstable_features: false,
             wait_to_build: None,
             show_warnings: true,
-            goto_def_racer_fallback: false,
             clear_env_rust_log: true,
             build_on_save: false,
             use_crate_blacklist: true,

